kv/kvnemesis: TestKVNemesisMultiNode failed

kv/kvnemesis.TestKVNemesisMultiNode [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10284115?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10284115?buildTab=artifacts#/) on release-22.2 @ [afe2d46600257f6c022ce89168bebdc0023c8215](https://github.com/cockroachdb/cockroach/commits/afe2d46600257f6c022ce89168bebdc0023c8215):


```
          | 	github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:96
          | github.com/cockroachdb/cockroach/pkg/rpc.bindUnaryServerInterceptorToHandler.func1
          | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:795
          | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func3
          | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:263
          | github.com/cockroachdb/cockroach/pkg/rpc.bindUnaryServerInterceptorToHandler.func1
          | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:795
          | github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor
          | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/auth.go:73
          | github.com/cockroachdb/cockroach/pkg/rpc.bindUnaryServerInterceptorToHandler.func1
          | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:795
          | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1.1
          | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:232
          | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr
          | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:341
          | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1
          | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:230
          | github.com/cockroachdb/cockroach/pkg/rpc.bindUnaryServerInterceptorToHandler.func1
          | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:795
          | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.func2
          | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:734
          | github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor.ClientInterceptor.func2
          | 	github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:226
          | github.com/cockroachdb/cockroach/pkg/rpc.getChainUnaryInvoker.func1
          | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:879
          | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.func3
          | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:764
          | github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.Batch
          | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:887
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).sendBatch
          | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport.go:211
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).SendNext
          | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport.go:190
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendToReplicas
          | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:2079
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendPartialBatch
          | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1623
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).divideAndSendBatchToRanges
          | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1221
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).Send
          | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:845
          | github.com/cockroachdb/cockroach/pkg/kv.(*CrossRangeTxnWrapperSender).Send
          | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:223
          | github.com/cockroachdb/cockroach/pkg/kv.(*DB).sendUsingSender
          | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:1016
        Wraps: (4) unable to find store 2 in range r56:/Table/100/"{405c460e"-857169cc"} [(n1,s1):1, (n4,s4):2, (n3,s3):3, next=5, gen=36, sticky=9223372036.854775807,2147483647]
        Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *errbase.opaqueWrapper (4) *errutil.leafError
    kvnemesis_test.go:113: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode897914563
--- FAIL: TestKVNemesisMultiNode (14.44s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #102411 kv/kvnemesis: TestKVNemesisMultiNode failed [C-test-failure O-robot T-kv X-duplicate todo-deprecate.branch-release-23.1.0]
- #101721 kv/kvnemesis: TestKVNemesisMultiNode failed [C-bug C-test-failure O-robot T-kv-replication branch-master]
- #100417 kv/kvnemesis: TestKVNemesisMultiNode failed [C-bug C-test-failure O-robot T-kv branch-release-23.1]
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisMultiNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvnemesis/validator.go
+++ b/pkg/kv/kvnemesis/validator.go
@@ -328,6 +328,35 @@ func (v *validator) getDeleteForKey(key string, txn *roachpb.Transaction) (stora
 	return storage.MVCCKey{}, false
 }
 
+func leaseTransferResultIsIgnorable(r Result) (ignore bool) {
+	err := errorFromResult(r)
+	if err == nil {
+		return false
+	}
+	return kvserver.IsLeaseTransferRejectedBecauseTargetMayNeedSnapshotError(err) ||
+		// Only VOTER (_FULL, _INCOMING, sometimes _OUTGOING) replicas can
+		// hold a range lease. Attempts to transfer to lease to any other
+		// replica type are rejected. See CheckCanReceiveLease.
+		resultIsErrorStr(r, `replica cannot hold lease`) ||
+		// Only replicas that are part of the range can be given
+		// the lease. This case is hit if a TransferLease op races
+		// with a ChangeReplicas op.
+		resultIsErrorStr(r, `replica not found in RangeDescriptor`) ||
+		// A lease transfer that races with a replica removal may find that
+		// the store it was targeting is no longer part of the range.
+		resultIsErrorStr(r, `unable to find store \d+ in range`) ||
+		// A lease transfer is not permitted while a range merge is in its
+		// critical phase.
+		resultIsErrorStr(r, `cannot transfer lease while merge in progress`) ||
+		// If the existing leaseholder has not yet heard about the transfer
+		// target's liveness record through gossip, it will return an error.
+		resultIsError(r, liveness.ErrRecordCacheMiss) ||
+		// Same as above, but matches cases where ErrRecordCacheMiss is
+		// passed through a LeaseRejectedError. This is necessary until
+		// LeaseRejectedErrors works with errors.Cause.
+		resultIsErrorStr(r, liveness.ErrRecordCacheMiss.Error())
+}
+
 func (v *validator) processOp(txnID *string, op Operation) {
 	switch t := op.GetValue().(type) {
 	case *GetOperation:
@@ -482,7 +511,7 @@ func (v *validator) processOp(txnID *string, op Operation) {
 			v.failIfError(op, t.Result)
 		}
 	case *ChangeReplicasOperation:
-		var ignore bool
+		ignore := leaseTransferResultIsIgnorable(t.Result)
 		if err := errorFromResult(t.Result); err != nil {
 			ignore = kvserver.IsRetriableReplicationChangeError(err) ||
 				kvserver.IsIllegalReplicationChangeError(err) ||
@@ -492,31 +521,7 @@ func (v *validator) processOp(txnID *string, op Operation) {
 			v.failIfError(op, t.Result)
 		}
 	case *TransferLeaseOperation:
-		var ignore bool
-		if err := errorFromResult(t.Result); err != nil {
-			ignore = kvserver.IsLeaseTransferRejectedBecauseTargetMayNeedSnapshotError(err) ||
-				// Only VOTER (_FULL, _INCOMING, sometimes _OUTGOING) replicas can
-				// hold a range lease. Attempts to transfer to lease to any other
-				// replica type are rejected. See CheckCanReceiveLease.
-				resultIsErrorStr(t.Result, `replica cannot hold lease`) ||
-				// Only replicas that are part of the range can be given
-				// the lease. This case is hit if a TransferLease op races
-				// with a ChangeReplicas op.
-				resultIsErrorStr(t.Result, `replica not found in RangeDescriptor`) ||
-				// A lease transfer that races with a replica removal may find that
-				// the store it was targeting is no longer part of the range.
-				resultIsErrorStr(t.Result, `unable to find store \d+ in range`) ||
-				// A lease transfer is not permitted while a range merge is in its
-				// critical phase.
-				resultIsErrorStr(t.Result, `cannot transfer lease while merge in progress`) ||
-				// If the existing leaseholder has not yet heard about the transfer
-				// target's liveness record through gossip, it will return an error.
-				resultIsError(t.Result, liveness.ErrRecordCacheMiss) ||
-				// Same as above, but matches cases where ErrRecordCacheMiss is
-				// passed through a LeaseRejectedError. This is necessary until
-				// LeaseRejectedErrors works with errors.Cause.
-				resultIsErrorStr(t.Result, liveness.ErrRecordCacheMiss.Error())
-		}
+		ignore := leaseTransferResultIsIgnorable(t.Result)
 		if !ignore {
 			v.failIfError(op, t.Result)
 		}

